File: contests/hc18/p4/p4_solution.cpp
#include <iostream>
#include <vector>
#include <string.h>
#include <algorithm>
#define MAXN 1000001

struct query_t {
    int l, r, k, index;
};

struct entry_t {
    int v, index;
};

int bit[MAXN];
int n;

bool entry_cmp(entry_t a, entry_t b) {
    return a.v < b.v;
}

bool query_cmp(query_t a, query_t b) {
    return a.k < b.k;
}

void update(int index, int value) {
    for(;index<=n; index += index&-index){
        bit[index]+=value;
    }
}

int query(int index) {
    int r = 0;
    for(;index > 0; index-=index&-index) {
        r += bit[index];
    }
    return r;
}

int main() {
    int q;
    std::cin >> n >> q;

    memset(bit, 0, sizeof(bit));

    std::vector<entry_t> entries;
    std::vector<query_t> queries;

    for (int i = 0; i < n; ++i) {
        int v;
        std::cin >> v;
        entries.push_back({v,i});
    }

    for (int i = 0; i < q; ++i) {
        int l, r, k;
        std::cin >> l >> r >> k;
        queries.push_back({l,r,k,i});
    }

    std::sort(entries.begin(), entries.end(), entry_cmp);
    std::sort(queries.begin(), queries.end(), query_cmp);

    int current = 0;
    std::vector<int> result(q);

    for (int i = 0; i < q; ++i) {
        while(current < n && entries[current].v <= queries[i].k) {
            update(entries[current].index+1,1);
            current++;
        }

        result[queries[i].index]=query(queries[i].r+1)-query(queries[i].l);
    }

    for (auto& r : result) {
        std::cout << r << "\n";
    }
}

File: contests/hc18/p4/p4_solution.py
import sys
raw_input = sys.stdin.readline

class Query(object):
    def __init__(self, left, right, k, index):
        self.left = left
        self.right = right
        self.k = k
        self.index = index

class Entry(object):
    def __init__(self, value, index):
        self.value = value
        self.index = index

n,q = map(int,raw_input().split())
bit = [0]*(n+1)

def update(index, value):
    while index <= n:
        bit[index] += value
        index += index & -index

def query(index):
    result = 0
    while index > 0:
        result += bit[index]
        index -= index & -index

    return result

costs = [int(i) for i in raw_input().split()]
entries = []
for i in range(n):
    entries.append(Entry(costs[i], i))

queries = []
for i in range(q):
    l,r,k = map(int,raw_input().split())
    queries.append(Query(l, r, k, i))

entries.sort(key=lambda x: x.value)
queries.sort(key=lambda x: x.k)

current = 0
results = [0]*q

for i in range(q):
    while current < n and entries[current].value <= queries[i].k:
        update(entries[current].index + 1, 1)
        current += 1
    
    results[queries[i].index] = query(queries[i].right+1) - query(queries[i].left)

for i in range(q):
    print(results[i])